modules/lib/products.js

```diff
@@ -37,31 +37,42 @@ module.exports.CLIENTS = {
   },
 }
 
-module.exports.getSetLock = (dbName) => async ({ user_id, channel, product }) => {
+const getKey = ({ channel, product, timestamp, tsFloor = 1000 * 60 * 5 }) => {
+  let key = `${channel}::${product}`
+  if (tsFloor > 0) {
+    const rounded = new Date(Math.floor(timestamp.getTime() / tsFloor) * tsFloor)
+    // in the form of YYYYMMDDHHmm 202110041615
+    key += `::${rounded.toISOString().split('.')[0].replace(/[^0-9]/g, '')}`
+  }
+  return key
+}
+
+module.exports.getSetLock = (dbName) => async ({ user_id, channel, product, hard = true, tsFloor = 1000 * 60 * 5 }) => {
   if (!process.env.DETA_KEY) {
     return true
   }
   const deta = Deta(process.env.DETA_KEY)
   const db = deta.Base(dbName)
-  const key = `${channel}::${product}`
+  const timestamp = new Date()
+  const key = getKey({ channel, product, timestamp, tsFloor })
+  const payload = { timestamp, user_id, channel, product }
+  if (hard) { // use deta.Base.insert to force error if already exists
+    return await db.insert(payload, key).catch(() => ({ ...payload, hard, locked: true }))
+  }
   const exists = await db.get(key)
   if (exists) {
     return { ...exists, locked: true }
   }
-  return await db.put({
-    timestamp: new Date(),
-    user_id,
-    channel,
-    product,
-  }, key)
+  return await db.put(payload, key)
 }
 
-module.exports.releaseLock = (dbName) => ({ channel, product }) => {
+module.exports.releaseLock = (dbName) => ({ channel, product, tsFloor = 1000 * 60 * 5 }) => {
```

Comment:
Maybe this could alternatively take `key` as input (sourced from `getSetLock`). That way you would always release the exact same lock you were given (e.g. if `getSetLock` is not in the same time interval as `releaseLock`) which would also help with cleaning up stale keys.

Reply:
indeed, `key` would be more flexible to the users of the getSet/release lock users

Reply:
Was thinking more of having the `getSet` function return the lock key as part of the payload instead of generating the key again at release time (which might end up being different as it is time-based).
```getSet(...args) -> { ...payload, key } -> releaseLock(key)```.

Reply:
i see. the "soft" version (with `get` to check existence) would contain that key. will add that to when it's `hard`

generating the key again in release time here is intended as there are 2 stateless steps involved (the slack command handler and the asynchronous worker)